save-preprocessor/src/main/kotlin/org/cqfn/save/preprocessor/utils/GitUtil.kt

```diff
@@ -5,25 +5,161 @@
 package org.cqfn.save.preprocessor.utils
 
 import org.cqfn.save.entities.GitDto
+import org.eclipse.jgit.api.CreateBranchCommand
 import org.eclipse.jgit.api.Git
+import org.eclipse.jgit.api.MergeCommand
+import org.eclipse.jgit.api.ResetCommand
+import org.eclipse.jgit.api.errors.GitAPIException
+import org.eclipse.jgit.api.errors.RefAlreadyExistsException
+import org.eclipse.jgit.api.errors.RefNotAdvertisedException
+import org.eclipse.jgit.api.errors.RefNotFoundException
+import org.eclipse.jgit.lib.Constants
+import org.eclipse.jgit.lib.Ref
 import org.eclipse.jgit.transport.CredentialsProvider
 import org.eclipse.jgit.transport.UsernamePasswordCredentialsProvider
+import org.slf4j.LoggerFactory
 import java.io.File
 
+private val log = LoggerFactory.getLogger(object {}.javaClass.enclosingClass::class.java)
+
 /**
  * @param gitDto
  * @param tmpDir
+ * @param branchOrCommit
  * @return jGit git entity
  */
-fun cloneFromGit(gitDto: GitDto, tmpDir: File): Git? {
+fun pullOrCloneProjectWithSpecificBranch(gitDto: GitDto, tmpDir: File, branchOrCommit: String?): Git? {
     val userCredentials = if (gitDto.username != null && gitDto.password != null) {
         UsernamePasswordCredentialsProvider(gitDto.username, gitDto.password)
     } else {
         CredentialsProvider.getDefault()
     }
+
+    if (tmpDir.exists() && !tmpDir.list().isNullOrEmpty()) {
+        pullProject(gitDto, tmpDir, userCredentials, branchOrCommit)?.let {
+            return it
+        }
+        // Pull was unsuccessful, clean directory before cloning
+        deleteDirectory(tmpDir)
+        generateDirectory(tmpDir)
+    }
+    log.info("Starting clone project ${gitDto.url} into the $tmpDir")
     return Git.cloneRepository()
         .setURI(gitDto.url)
         .setCredentialsProvider(userCredentials)
         .setDirectory(tmpDir)
+        .call().also { git ->
+            switchBranch(git, gitDto.url, branchOrCommit)
+        }
+}
+
+/**
+ * @param gitDto
+ * @param tmpDir
+ * @param userCredentials
+ * @param branchOrCommit
+ * @return jGit git entity
+ */
+fun pullProject(
+    gitDto: GitDto,
+    tmpDir: File,
+    userCredentials: CredentialsProvider?,
+    branchOrCommit: String?
+): Git? {
+    log.info("Starting pull project ${gitDto.url} into the $tmpDir")
+    val git = Git.open(tmpDir)
```

Comment:
Do operations on this `git` need to be wrapped in `use`?

Reply:
Basically, if all is ok, this method return `git`, which is transfers to the `pullOrCloneProjectWithSpecificBranch`, which wrapped in `use` in places, where do we call this function

However, when we return `null`, it will be better to close `git`. despite the fact, that I didn't met any problems, when tested it